persist: admin tool to manually compact a shard

danhhz opened the pull request:
Persist shards currently only perform compactions that they directly caused by inserting a batch. Any others are assumed to be done by someone else. Among other things, this doesn't backfill compactions when we change our selection criteria (such as when we started considering batch parts). This is one of the things we'll be revisiting in Persist Compaction 2.0, but until then add an admin tool we can use to manually fixup individual shards.

Touches #16161

### Motivation

  * This PR adds a known-desirable feature.

### Tips for reviewer

Left an open question about BuildInfo. Time to add persist to the list of things that get the version set?

Manually tested this against ./bin/environmentd, but we'll probably want to do a bit more testing on a staging cluster before we point it at anything real.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

pH14 commented on src/persist-client/examples/admin.rs:
```diff
@@ -0,0 +1,91 @@
+pub struct AdminArgs {
+    #[clap(subcommand)]
+    command: Command,
+}
```
let's make `read_only` one of the stock AdminArgs arguments, and have it default to `true`

danhhz replied:
you're about to find out lovely things about clap (bool that defaults to true is not possible 🚮). love the idea though, done!

pH14 commented on src/persist-client/src/inspect.rs:
```diff
@@ -324,6 +333,139 @@ pub async fn unreferenced_blobs(
+                start.elapsed(),
+            );
+            let apply_res = machine
+                .merge_res(&FueledMergeRes { output: res.output })
```
let's wrap this in the `if !read_only` check we plumb through

pH14 replied:
👍 what you have here is what I had in mind, it's handy to have a no-op mode just to double check the work that would be done

pH14 commented on src/persist-client/examples/inspect.rs:
```diff
@@ -7,14 +7,14 @@
 use std::str::FromStr;
 
-/// Commands for inspecting current persist state
+/// Commands for read-only inspection of persist state
```
oh nice, I like separating this out 🙌

pH14 commented on src/persist-client/src/inspect.rs:
```diff
@@ -324,6 +333,139 @@ pub async fn unreferenced_blobs(
     Ok(unreferenced_blobs)
 }
 
+/// Manually completes all fueled compactions in a shard.
```
since we split out `inspect` and `admin` cmds it seems odd to have this in `inspect.rs`

danhhz replied:
I was trying to keep K V and KVTD_CODECS from leaking out of the mod 🙈 . done!

pH14 commented on src/persist-client/examples/admin.rs:
```diff
@@ -0,0 +1,91 @@
+                sha: "0000000000000000000000000000000000000000",
+                time: "",
+            };
+            let cfg = PersistConfig::new(&build_info, SYSTEM_TIME.clone());
```
we can plumb through an arg for the memory bound too, since we can use (almost) the entire pod memory limit we choose for exclusively compaction

danhhz replied:
ran out of time for this one

pH14 reviewed (commented):
🎉

danhhz commented:
my plane just started its descent so it's highly likely I won't get back to this before monday. @pH14 if you're around tomorrow and there is time pressure to get this merged, feel free to take it over. otherwise, I'll pick it up first thing

rjnn commented:
I think I have a long running load generator source in my environment (4+ weeks) which likely means there's a lot of uncompacted state). Happy to volunteer my environment for testing.

pH14 commented on src/persist-client/examples/inspect.rs:
```diff
@@ -82,15 +82,15 @@ pub struct StateArgs {
     ///
     #[clap(long, verbatim_doc_comment)]
-    consensus_uri: String,
+    pub(crate) consensus_uri: String,
```
while you're here, would you mind creating equivalent env vars to set the consensus and blob uris? we'll want env vars so we're not passing around secrets in the Kube pod spec

danhhz commented on src/persist-client/examples/admin.rs:
```diff
@@ -0,0 +1,91 @@
+    match command.command {
+        Command::ForceCompaction(args) => {
+            let shard_id = ShardId::from_str(&args.shard_id).expect("invalid shard id");
+            // WIP hook this up to a real build_info
```
@benesch any objections to putting persist-client in the ./bin/bump-version script?

danhhz reviewed (commented):
okay this should be ready to go!

> I think I have a long running load generator source in my environment (4+ weeks) which likely means there's a lot of uncompacted state). Happy to volunteer my environment for testing.

@rjnn you indeed have an affected shard (sfc96b238 - edac41b73 has >40k parts) so we'll likely take you up on this!

danhhz commented:
TFTR! squashed and updated the persist-client version to 0.34.0-dev (it moved yesterday). waiting on @benesch's opinion on adding persist-client to the list of things versions by bin/bump-version before merging

danhhz commented:
Added a last minute thought I had to bail if the version of state it fetches at startup isn't exactly the version being used by the tool. This should help a ton with preventing accidental badness via misuse
